#!/usr/bin/env python
# Copyright (c) 2008 Qtrac Ltd. All rights reserved.
# This program or module is free software: you can redistribute it and/or
# modify it under the terms of the GNU General Public License as published
# by the Free Software Foundation, either version 2 of the License, or
# version 3 of the License, or (at your option) any later version. It is
# provided for educational purposes and is distributed in the hope that
# it will be useful, but WITHOUT ANY WARRANTY; without even the implied
# warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See
# the GNU General Public License for more details.

from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals
from future_builtins import *

from PyQt4.QtCore import (QVariant, Qt, SIGNAL)
from PyQt4.QtGui import (QApplication, QBrush, QColorDialog, QDialog,
        QPainter, QPixmap)
import ui_newimagedlg


class NewImageDlg(QDialog, ui_newimagedlg.Ui_NewImageDlg):

    def __init__(self, parent=None):
        super(NewImageDlg, self).__init__(parent)
        self.setupUi(self)

        self.color = Qt.red
        for value, text in (
                (Qt.SolidPattern, self.tr("Solid")),
                (Qt.Dense1Pattern, self.tr("Dense #1")),
                (Qt.Dense2Pattern, self.tr("Dense #2")),
                (Qt.Dense3Pattern, self.tr("Dense #3")),
                (Qt.Dense4Pattern, self.tr("Dense #4")),
                (Qt.Dense5Pattern, self.tr("Dense #5")),
                (Qt.Dense6Pattern, self.tr("Dense #6")),
                (Qt.Dense7Pattern, self.tr("Dense #7")),
                (Qt.HorPattern, self.tr("Horizontal")),
                (Qt.VerPattern, self.tr("Vertical")),
                (Qt.CrossPattern, self.tr("Cross")),
                (Qt.BDiagPattern, self.tr("Backward Diagonal")),
                (Qt.FDiagPattern, self.tr("Forward Diagonal")),
                (Qt.DiagCrossPattern, self.tr("Diagonal Cross"))):
            self.brushComboBox.addItem(text, QVariant(value))

        self.connect(self.colorButton, SIGNAL("clicked()"),
                     self.getColor)
        self.connect(self.brushComboBox, SIGNAL("activated(int)"),
                     self.setColor)
        self.setColor()
        self.widthSpinBox.setFocus()


    def getColor(self):
        color = QColorDialog.getColor(Qt.black, self)
        if color.isValid():
            self.color = color
            self.setColor()


    def setColor(self):
        pixmap = self._makePixmap(60, 30)
        self.colorLabel.setPixmap(pixmap)


    def image(self):
        pixmap = self._makePixmap(self.widthSpinBox.value(),
                                  self.heightSpinBox.value())
        return QPixmap.toImage(pixmap)


    def _makePixmap(self, width, height):
        pixmap = QPixmap(width, height)
        style = self.brushComboBox.itemData(
                        self.brushComboBox.currentIndex()).toInt()[0]
        brush = QBrush(self.color, Qt.BrushStyle(style))
        painter = QPainter(pixmap)
        painter.fillRect(pixmap.rect(), Qt.white)
        painter.fillRect(pixmap.rect(), brush)
        return pixmap


if __name__ == "__main__":
    import sys

    app = QApplication(sys.argv)
    form = NewImageDlg()
    form.show()
    app.exec_()

